File: food.py
import random
from dataclasses import dataclass
from enum import IntEnum


class FoodType(IntEnum):
    APPLE = 0
    BANANA = 1
    GOLDEN_APPLE = 2


FOOD_POINTS: dict[FoodType, int] = {
    FoodType.APPLE: 25,
    FoodType.BANANA: 50,
    FoodType.GOLDEN_APPLE: 200,
}

REGULAR_INTERVAL = 5.0
GOLDEN_INTERVAL = 30.0
MAX_REGULAR_FOOD = 3


@dataclass
class FoodItem:
    col: int
    row: int
    food_type: FoodType
    spawn_time: float


class FoodManager:
    def __init__(self, grid_cols: int, grid_rows: int) -> None:
        self._cols = grid_cols
        self._rows = grid_rows
        self.items: list[FoodItem] = []
        self.golden: FoodItem | None = None
        self._last_regular_spawn: float = 0.0
        self._last_golden_spawn: float = 0.0

    def reset(self, now: float) -> None:
        self.items.clear()
        self.golden = None
        self._last_regular_spawn = now
        self._last_golden_spawn = now

    def update(self, now: float, snake_body: list[tuple[int, int]],
               snake_head_next: tuple[int, int]) -> None:
        existing = [(i.col, i.row) for i in self.items]
        if self.golden:
            existing.append((self.golden.col, self.golden.row))

        # Regular food
        if len(self.items) == 0:
            self._spawn_regular(now, snake_body, snake_head_next, existing)
        elif now - self._last_regular_spawn >= REGULAR_INTERVAL and len(self.items) < MAX_REGULAR_FOOD:
            self._spawn_regular(now, snake_body, snake_head_next, existing)

        # Golden apple
        if self.golden is None and now - self._last_golden_spawn >= GOLDEN_INTERVAL:
            existing2 = [(i.col, i.row) for i in self.items]
            cell = self._find_valid_cell(snake_body, snake_head_next, existing2)
            if cell:
                self.golden = FoodItem(cell[0], cell[1], FoodType.GOLDEN_APPLE, now)
                self._last_golden_spawn = now

    def try_eat(self, col: int, row: int) -> FoodItem | None:
        if self.golden and self.golden.col == col and self.golden.row == row:
            item = self.golden
            self.golden = None
            return item
        for i, item in enumerate(self.items):
            if item.col == col and item.row == row:
                self.items.pop(i)
                return item
        return None

    def get_all_items(self) -> list[FoodItem]:
        return list(self.items) + ([self.golden] if self.golden else [])

    def offset_timers(self, delta: float) -> None:
        self._last_regular_spawn += delta
        self._last_golden_spawn += delta

    def _spawn_regular(self, now: float, snake_body: list[tuple[int, int]],
                       forbidden: tuple[int, int],
                       existing_food: list[tuple[int, int]]) -> None:
        cell = self._find_valid_cell(snake_body, forbidden, existing_food)
        if cell:
            food_type = random.choice([FoodType.APPLE, FoodType.BANANA])
            self.items.append(FoodItem(cell[0], cell[1], food_type, now))
            self._last_regular_spawn = now

    def _find_valid_cell(self, snake_body: list[tuple[int, int]],
                         forbidden: tuple[int, int],
                         existing_food: list[tuple[int, int]]) -> tuple[int, int] | None:
        occupied = set(snake_body) | {forbidden} | set(existing_food)
        free = [(c, r) for c in range(self._cols) for r in range(self._rows)
                if (c, r) not in occupied]
        return random.choice(free) if free else None

File: src/game/food.ts
import type { Cell } from './types';
import type { Rng } from './rng';

export enum FoodType {
  APPLE = 'apple',
  BANANA = 'banana',
  GOLDEN_APPLE = 'golden_apple',
}

export const FOOD_POINTS: Readonly<Record<FoodType, number>> = {
  [FoodType.APPLE]: 25,
  [FoodType.BANANA]: 50,
  [FoodType.GOLDEN_APPLE]: 200,
};

export interface FoodItem {
  col: number;
  row: number;
  foodType: FoodType;
}

const REGULAR_SPAWN_INTERVAL = 5.0;
const GOLDEN_SPAWN_INTERVAL = 30.0;
const MAX_REGULAR = 3;

export class FoodManager {
  private items: FoodItem[] = [];
  private lastRegularSpawn = 0;
  private lastGoldenSpawn = 0;
  private readonly cols: number;
  private readonly rows: number;
  private readonly rng: Rng;

  constructor(gridCols: number, gridRows: number, rng: Rng) {
    this.cols = gridCols;
    this.rows = gridRows;
    this.rng = rng;
  }

  reset(now: number): void {
    this.items = [];
    this.lastRegularSpawn = now;
    this.lastGoldenSpawn = now;
  }

  update(now: number, snakeBody: Cell[], cellInFront: Cell): void {
    const regulars = this.items.filter(i => i.foodType !== FoodType.GOLDEN_APPLE);

    if (regulars.length === 0) {
      const cell = this.pickCell(snakeBody, cellInFront);
      if (cell) {
        const type = this.rng.next() < 0.5 ? FoodType.APPLE : FoodType.BANANA;
        this.items.push({ col: cell.col, row: cell.row, foodType: type });
        this.lastRegularSpawn = now;
      }
    } else if (regulars.length < MAX_REGULAR && now - this.lastRegularSpawn >= REGULAR_SPAWN_INTERVAL) {
      const cell = this.pickCell(snakeBody, cellInFront);
      if (cell) {
        const type = this.rng.next() < 0.5 ? FoodType.APPLE : FoodType.BANANA;
        this.items.push({ col: cell.col, row: cell.row, foodType: type });
        this.lastRegularSpawn = now;
      }
    }

    const hasGolden = this.items.some(i => i.foodType === FoodType.GOLDEN_APPLE);
    if (!hasGolden && now - this.lastGoldenSpawn >= GOLDEN_SPAWN_INTERVAL) {
      const cell = this.pickCell(snakeBody, cellInFront);
      if (cell) {
        this.items.push({ col: cell.col, row: cell.row, foodType: FoodType.GOLDEN_APPLE });
        this.lastGoldenSpawn = now;
      }
    }
  }

  tryEat(col: number, row: number): FoodItem | null {
    const idx = this.items.findIndex(i => i.col === col && i.row === row);
    if (idx === -1) return null;
    // biome-ignore lint/style/noNonNullAssertion: idx !== -1 guarantees splice returns ≥1 element
    return this.items.splice(idx, 1)[0]!;
  }

  getAllItems(): FoodItem[] {
    return [...this.items];
  }

  offsetTimers(delta: number): void {
    this.lastRegularSpawn += delta;
    this.lastGoldenSpawn += delta;
  }

  private pickCell(snakeBody: Cell[], cellInFront: Cell): Cell | null {
    const occupied = new Set<string>(snakeBody.map(c => `${c.col},${c.row}`));
    occupied.add(`${cellInFront.col},${cellInFront.row}`);
    for (const item of this.items) {
      occupied.add(`${item.col},${item.row}`);
    }

    for (let attempt = 0; attempt < 100; attempt++) {
      const col = this.rng.nextInt(0, this.cols - 1);
      const row = this.rng.nextInt(0, this.rows - 1);
      if (!occupied.has(`${col},${row}`)) {
        return { col, row };
      }
    }
    return null;
  }
}
